Add initial demo-driven design demos

hlandau opened the pull request:
Previous discussion: https://github.com/hlandau/openssl-ddd/pull/1
See also #17939.

For discussion of the "demo-driven design" demos and permanent inclusion of these demos into the OpenSSL tree for future reference.

This comment will be updated to link to the blogpost when it's published.

mattcaswell commented on ddd/ddd-05-mem-nonblocking.c:
```diff
@@ -0,0 +1,408 @@
+                    case EAGAIN:
+                        goto stop;
+                    default:
+                        fprintf(stderr, "error on read: %d\n", errno);
```
Needs to handle EOF which is not an error in itself

mattcaswell commented on ddd/ddd-01-conn-blocking.c:
```diff
@@ -0,0 +1,161 @@
+        goto fail;
+    }
+
+    b = new_conn(ctx, "www.example.com:443");
```
Should we use something other than example.com? I'm thinking that this will eventually make its way into CIs and relying on a 3rd party domain may not be the way to go.

mattcaswell commented on ddd/Makefile:
```diff
@@ -0,0 +1,10 @@
+
+ddd-%: ddd-%.c
+	$(CC) -O3 -g -o "$@" "$<" -lcrypto -lssl
+
```
Please add a clean target

hlandau commented on ddd/README.md:
```diff
@@ -0,0 +1,87 @@
+Demo-Driven Design
+==================
+
+The OpenSSL project from time to time must evolve its public API surface in
```
Added a scope section clarifying the limited scope.

hlandau commented:
Updated:
- Added notes on scope to README.
- Added clean target.
- Warnings.
- Added note about `SSL_CERT_DIR` to README.
- Used `www.openssl.org:443` as endpoint.
- Remove unnecessary buffering in nonblocking RX code.
- Handle EOF in ddd-05.

hlandau commented:
Unclear to me how one is supposed to avoid long lines in Markdown tables. I've exempted it from the mdl config.

hlandau commented:
@mattcaswell Saw your notes on `SSL_set_connect_state`. It is interesting that removing `SSL_set_connect_state` worked for you. Is there some kind of undefined state we have here we should be explicitly checking for?

mattcaswell commented:
> It is interesting that removing SSL_set_connect_state worked for you.

This was (partly) user error. Removing it didn't actually report any errors, but it didn't come back with a response either. At the time I hadn't set `SSL_CERT_DIR` so I think that might have also masked things.

hlandau commented:
My personal proposed diffs against demos 1 through 4 are now available for viewing here: https://github.com/hlandau/openssl/compare/ddd-diff-base...hlandau:ddd-diff

Feedback welcomed, and since I expect this to be quite a collaborative process, if other people have very different ideas for what the diffs should look like, I'd be interested to see.

hlandau commented:
The proposed diff for demo 5 is now available here:
https://github.com/hlandau/openssl/compare/ddd-diff-base...hlandau:ddd-diff

Each demo has a Markdown file discussing the design issues and resultant changes. You can read these discussions inline in the above diffs, or here:

https://github.com/hlandau/openssl/tree/ddd-diff/ddd

hlandau commented:
I've added a demo 6, which demonstrates use of a memory BIO with libuv, the asynchronous I/O library. I believe this captures valuable design information that is not necessarily obvious in demo 5.

hlandau commented:
Demo 2 has been chosen to exemplify use of BIO_f_buffer. Diffs and the discussion have been updated.

t8m commented:
> The proposed diff for demo 5 is now available here: [hlandau/openssl@ddd-diff-base...hlandau:ddd-diff](https://github.com/hlandau/openssl/compare/ddd-diff-base...hlandau:ddd-diff)
> 
> Each demo has a Markdown file discussing the design issues and resultant changes. You can read these discussions inline in the above diffs, or here:
> 
> https://github.com/hlandau/openssl/tree/ddd-diff/ddd

Could you create a pull request on your own fork at least so I can add comments to the diff? I have some questions and concerns with some of the diffs.

hlandau commented:
@t8m Done. https://github.com/hlandau/openssl/pull/1

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

paulidale commented:
Also approved if the directories are moved somewhere a little better (e.g. doc/design)

t8m commented:
> Also approved if the directories are moved somewhere a little better (e.g. doc/design)

Yeah, please move it there -> doc/design/ddd

hlandau commented:
Moved.

t8m commented:
Assuming @paulidale is ok with the move.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to master branch. Thank you.
